File: runtime/apps/tui/src/types/event.ts
import type { Message, MessagePart, Session } from "./session.js";
import type { PermissionRequest, QuestionRequest } from "./permission.js";

export interface GatewayEventEnvelope {
  directory?: string;
  payload?: GatewayEventPayload;
}

export interface SessionEventProperties {
  sessionID: string;
  info: Session;
}

export interface SessionStatusEventProperties {
  sessionID: string;
  updatedAt: number;
  status: unknown;
  context_tokens: Session["context_tokens"];
  usage?: Session["usage"];
}

export interface MessageUpdatedEventProperties {
  sessionID: string;
  info: Message;
}

export interface MessageRemovedEventProperties {
  sessionID: string;
  messageID: string;
}

export interface MessagePartDeltaEventProperties {
  sessionID: string;
  messageID: string;
  partID: string;
  createdAt: number;
  updatedAt: number;
  field: string;
  delta: string;
}

export interface MessagePartUpdatedEventProperties {
  sessionID: string;
  createdAt: number;
  updatedAt: number;
  part: MessagePart;
}

export interface CommandUpdatedEventProperties {
  sessionID: string;
  messageID: string;
  partID: string;
  runtimeID: string;
  commandRunID: string;
  commandID: string;
  providerToolCallID?: string | null;
  commandIndex?: number | null;
  eventSeq?: number | null;
  status: string;
  command?: unknown;
  result?: unknown;
  createdAt: number;
  updatedAt: number;
}

export function commandStatusIsRunning(status: string): boolean {
  return /^(running|pending|busy|in[_ -]?progress|executing|started)$/iu.test(status.trim());
}

export type GatewayEventPayload =
  | { type: "server.connected"; properties: Record<string, unknown> }
  | { type: "session.created"; properties: SessionEventProperties }
  | { type: "session.updated"; properties: SessionEventProperties }
  | { type: "session.deleted"; properties: SessionEventProperties }
  | { type: "session.status"; properties: SessionStatusEventProperties }
  | { type: "message.updated"; properties: MessageUpdatedEventProperties }
  | { type: "message.removed"; properties: MessageRemovedEventProperties }
  | { type: "message.part.delta"; properties: MessagePartDeltaEventProperties }
  | { type: "message.part.updated"; properties: MessagePartUpdatedEventProperties }
  | { type: "command.updated"; properties: CommandUpdatedEventProperties }
  | { type: "permission.asked"; properties: Record<string, unknown> }
  | { type: "permission.replied"; properties: Record<string, unknown> }
  | { type: "question.asked"; properties: Record<string, unknown> }
  | { type: "question.replied"; properties: Record<string, unknown> }
  | { type: "question.rejected"; properties: Record<string, unknown> }
  | { type: string; properties?: Record<string, unknown> };

export interface NormalizedEvent {
  type: string;
  directory: string;
  sessionID?: string;
  messageID?: string;
  role?: Message["role"];
  partID?: string;
  status?: string;
  text?: string;
  tool?: string;
  commandID?: string;
  permission?: PermissionRequest;
  question?: QuestionRequest;
  raw: GatewayEventEnvelope;
}

export function eventSessionID(payload: GatewayEventPayload | undefined): string | undefined {
  const properties = payload?.properties as Record<string, unknown> | undefined;
  const direct = properties?.sessionID as string | undefined;
  if (direct) return direct;
  const info = properties?.info as Record<string, unknown> | undefined;
  const infoSession = info?.sessionID as string | undefined;
  if (infoSession) return infoSession;
  const part = properties?.part as Record<string, unknown> | undefined;
  const partSession = part?.sessionID as string | undefined;
  if (partSession) return partSession;
  if (
    payload?.type !== "permission.asked" &&
    payload?.type !== "permission.replied" &&
    payload?.type !== "question.asked" &&
    payload?.type !== "question.replied" &&
    payload?.type !== "question.rejected"
  ) {
    return undefined;
  }
  const request = (properties?.permission ??
    properties?.question ??
    properties?.request ??
    properties) as Record<string, unknown> | undefined;
  return request?.sessionID as string | undefined;
}

File: runtime/crates/gateway/src/contracts/event.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use super::{Message, Project, Session, SessionContextTokens, SessionUsage};

// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum GlobalEvent {
    #[serde(rename = "server.connected")]
    ServerConnected {
        properties: HashMap<String, serde_json::Value>,
    },
    #[serde(rename = "server.instance.disposed")]
    ServerInstanceDisposed {
        properties: InstanceDisposedProperties,
    },
    #[serde(rename = "project.updated")]
    ProjectUpdated { properties: Project },
    #[serde(rename = "session.created")]
    SessionCreated {
        properties: SessionCreatedProperties,
    },
    #[serde(rename = "session.updated")]
    SessionUpdated {
        properties: SessionUpdatedProperties,
    },
    #[serde(rename = "session.deleted")]
    SessionDeleted {
        properties: SessionDeletedProperties,
    },
    #[serde(rename = "session.status")]
    SessionStatus { properties: SessionStatusProperties },
    #[serde(rename = "message.updated")]
    MessageUpdated {
        properties: MessageUpdatedProperties,
    },
    #[serde(rename = "message.removed")]
    MessageRemoved {
        properties: MessageRemovedProperties,
    },
    #[serde(rename = "message.part.delta")]
    MessagePartDelta {
        properties: MessagePartDeltaProperties,
    },
    #[serde(rename = "message.part.updated")]
    MessagePartUpdated {
        properties: MessagePartUpdatedProperties,
    },
    #[serde(rename = "command.updated")]
    CommandUpdated {
        properties: CommandUpdatedProperties,
    },
    #[serde(rename = "todo.updated")]
    TodoUpdated { properties: serde_json::Value },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InstanceDisposedProperties {
    pub directory: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionCreatedProperties {
    #[serde(rename = "sessionID")]
    pub session_id: String,
    pub info: Session,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionUpdatedProperties {
    #[serde(rename = "sessionID")]
    pub session_id: String,
    pub info: Session,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionDeletedProperties {
    #[serde(rename = "sessionID")]
    pub session_id: String,
    pub info: Session,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionStatusProperties {
    #[serde(rename = "sessionID")]
    pub session_id: String,
    #[serde(rename = "updatedAt")]
    pub updated_at: i64,
    pub status: serde_json::Value,
    #[serde(default)]
    pub context_tokens: SessionContextTokens,
    #[serde(default)]
    pub usage: SessionUsage,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MessageUpdatedProperties {
    #[serde(rename = "sessionID")]
    pub session_id: String,
    pub info: Message,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MessageRemovedProperties {
    #[serde(rename = "sessionID")]
    pub session_id: String,
    #[serde(rename = "messageID")]
    pub message_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MessagePartDeltaProperties {
    #[serde(rename = "sessionID")]
    pub session_id: String,
    #[serde(rename = "messageID")]
    pub message_id: String,
    #[serde(rename = "partID")]
    pub part_id: String,
    #[serde(rename = "createdAt")]
    pub created_at: i64,
    #[serde(rename = "updatedAt")]
    pub updated_at: i64,
    pub field: String,
    pub delta: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MessagePartUpdatedProperties {
    #[serde(rename = "sessionID")]
    pub session_id: String,
    #[serde(rename = "createdAt")]
    pub created_at: i64,
    #[serde(rename = "updatedAt")]
    pub updated_at: i64,
    pub part: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommandUpdatedProperties {
    #[serde(rename = "sessionID")]
    pub session_id: String,
    #[serde(rename = "messageID")]
    pub message_id: String,
    #[serde(rename = "partID")]
    pub part_id: String,
    #[serde(rename = "runtimeID")]
    pub runtime_id: String,
    #[serde(rename = "commandRunID")]
    pub command_run_id: String,
    #[serde(rename = "commandID")]
    pub command_id: String,
    #[serde(rename = "providerToolCallID", default)]
    pub provider_tool_call_id: Option<String>,
    #[serde(rename = "commandIndex", default)]
    pub command_index: Option<u64>,
    #[serde(rename = "eventSeq", default)]
    pub event_seq: Option<i64>,
    pub status: String,
    #[serde(default)]
    pub command: serde_json::Value,
    #[serde(default)]
    pub result: serde_json::Value,
    #[serde(rename = "createdAt")]
    pub created_at: i64,
    #[serde(rename = "updatedAt")]
    pub updated_at: i64,
}

// Sync events (lighter weight)
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum SyncEvent {
    #[serde(rename = "sync.project.updated")]
    ProjectUpdated { properties: Project },
    #[serde(rename = "sync.session.updated")]
    SessionUpdated { properties: Box<Session> },
}
